### PR TITLE
feat: Implements DISTS metric

### PR DESCRIPTION
**Pull request by koreaygj**
## Summery

- custom Pooling: Replaces standard MaxPooling with L2Pooling (Hanning window + L2 norm) for smoother, shift-invariant downsampling.
- Feature Extraction: Utilizes a VGG16L2PoolExtractor to capture multi-stage features across 6 different layers (input through stage 5).
- Pretrained Optimization: Employs pretrained weights for both the VGG16 backbone and DISTS-specific parameters ($\alpha, \beta$), similar to the LPIPS metric approach.

### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Related to https://github.com/tracel-ai/burn/issues/4312 (Image quality metrics)

### Changes

**Matric**

The metric evaluates two primary components using learnable $\alpha$ (alpha) and $\beta$ (beta) weights for each channel
Structure Similarity: Based on mean values.

$$S_{struct} = \frac{2\mu_x\mu_y + c_1}{\mu_x^2 + \mu_y^2 + c_1}$$

Texture Similarity: Based on variance and covariance.

$$S_{text} = \frac{2\sigma_{xy} + c_2}{\sigma_x^2 + \sigma_y^2 + c_2}$$

Final Score: A weighted combination of structural and textural distances across all extracted layers.

### Testing

Test Case | Description | Result
-- | -- | --
Identical Image Distance | Confirms distance is $0$ when comparing an image to itself ($D(A, A) = 0$). | ✅
Distinct Image Distance | Confirms distance is $> 0$ for different images ($D(A, B) > 0$). | ✅
Symmetry Test | Verified that the order of inputs does not affect the result ($D(A, B) = D(B, A)$). | ✅
Batch Processing | Confirmed the model correctly handles 4D tensor batches. | ✅
Non-Reduction Mode | Verified that the metric can return individual scores per batch element. | ✅
Pretrained Weights | Successfully loaded and remapped VGG16 and DISTS-specific weights. | ✅
Module Display | Confirmed the model structure and parameters are correctly initialized. | ✅

### CODE REVIEW

**Comment by koreaygj** [comment]
Add githubmac-runner ci burn-train vision test for `LPIPS`, `DISTS`

File: crates/burn-train/src/metric/vision/dists/metric.rs [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+    #[config(default = true)]
+    pub normalize: bool,
+}
+
+impl DistsConfig {
+    /// Initialize a DISTS module with random weights.
+    pub fn init<B: Backend>(&self, device: &B::Device) -> Dists<B> {
+        let total_channels: usize = CHANNELS.iter().sum();
+
+        // Initialize alpha and beta with normal distribution (mean=0.1, std=0.01)
```
[the remark this answers is not in the document]
Comment (koreaygj) [reply crates/burn-train/src/metric/vision/dists/metric.rs]: resolved.

File: crates/burn-train/src/metric/vision/dists/weights.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    dists
+}
+
+/// Load DISTS trained alpha/beta weights.
+fn load_dists_weights<B: Backend>(mut dists: Dists<B>, cache_path: &PathBuf) -> Dists<B> {
+    let mut store = PytorchStore::from_file(cache_path)
+        .allow_partial(true)
+        .skip_enum_variants(true)
+        // Alpha and beta weights
```
[the remark this answers is not in the document]
Comment (koreaygj) [reply crates/burn-train/src/metric/vision/dists/weights.rs]: resolved.

File: crates/burn-train/src/metric/vision/dists/metric.rs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+            .mul(x.clone())
+            .mean_dim(2)
+            .squeeze_dim::<2>(2)
+            .sub(mean_x.clone().mul(mean_x.clone()));
+        let var_y = y
+            .clone()
+            .mul(y.clone())
+            .mean_dim(2)
+            .squeeze_dim::<2>(2)
+            .sub(mean_y.clone().mul(mean_y.clone()));
```
Comment (laggui) [inline crates/burn-train/src/metric/vision/dists/metric.rs]: Given the denominator for

```
(2*cov_xy + c2) / (var_x + var_y + c2)
```

we might want to clamp variances at 0 for good measure.

File: crates/burn-train/src/metric/vision/dists/metric.rs [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+            .div(var_x.add(var_y).add(c2));
+
+        // Convert similarity to distance: 1 - similarity
+        let structure_dist = Tensor::<B, 2>::ones([batch, channels], &device).sub(structure_sim);
+        let texture_dist = Tensor::<B, 2>::ones([batch, channels], &device).sub(texture_sim);
+
+        // Apply weights: [batch, channels] * [channels] -> [batch, channels]
+        // Then sum over channels -> [batch]
+        let alpha_expanded = alpha.unsqueeze_dim::<2>(0).repeat_dim(0, batch);
+        let beta_expanded = beta.unsqueeze_dim::<2>(0).repeat_dim(0, batch);
```
Comment (laggui) [inline crates/burn-train/src/metric/vision/dists/metric.rs]: Instead of repeating the dims we should rely on broadcasting, which doesn't allocate new memory

File: crates/burn-train/src/metric/vision/dists/metric.rs [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+    fn preprocess(
+        &self,
+        input: Tensor<B, 4>,
+        target: Tensor<B, 4>,
+    ) -> (Tensor<B, 4>, Tensor<B, 4>) {
+        if self.normalize {
+            // Normalize from [0, 1] to [-1, 1]
+            let input = input.mul_scalar(2.0).sub_scalar(1.0);
+            let target = target.mul_scalar(2.0).sub_scalar(1.0);
+            (input, target)
```
Comment (laggui) [inline crates/burn-train/src/metric/vision/dists/metric.rs]: When using imagenet pre-trained weights, shouldn't normalization use the appropriate mean/std (`[0.485, 0.456, 0.406]` / `[0.229, 0.224, 0.225]`)?

File: crates/burn-train/src/metric/vision/dists/metric.rs (hunk 3 above)
Comment (koreaygj) [reply crates/burn-train/src/metric/vision/dists/metric.rs]: Good Point I applied

File: crates/burn-train/src/metric/vision/dists/metric.rs (hunk 4 above)
Comment (koreaygj) [reply crates/burn-train/src/metric/vision/dists/metric.rs]: resolved.

File: crates/burn-train/src/metric/vision/dists/metric.rs (hunk 5 above)
Comment (koreaygj) [reply crates/burn-train/src/metric/vision/dists/metric.rs]: I missed 🥺... Thanks for finding that!

Comment (koreaygj) [reply crates/burn-train/src/metric/vision/dists/metric.rs]: Only constant images (e.g., all ones) caused numerical edge cases with ImageNet normalization. Fixed by using random images in tests.

**Comment by koreaygj** [comment]
All changes applied!🤩

File: crates/burn-train/src/metric/vision/dists/metric.rs [hunk 6]
```diff
@@ -266,36 +268,44 @@ impl<B: Backend> Dists<B> {
...
+
+    /// Apply ImageNet normalization to a tensor.
+    /// mean = [0.485, 0.456, 0.406], std = [0.229, 0.224, 0.225]
+    fn imagenet_normalize(x: Tensor<B, 4>) -> Tensor<B, 4> {
+        // Normalize each channel: (x - mean) / std
+        let r = x.clone().narrow(1, 0, 1).sub_scalar(0.485).div_scalar(0.229);
+        let g = x.clone().narrow(1, 1, 1).sub_scalar(0.456).div_scalar(0.224);
+        let b = x.narrow(1, 2, 1).sub_scalar(0.406).div_scalar(0.225);
+
+        Tensor::cat(vec![r, g, b], 1)
```
Comment (laggui) [inline crates/burn-train/src/metric/vision/dists/metric.rs]: I'd use a mean and std tensor instead of having to split each channel and concat the output.

```rust
let mean = Tensor::<B, 1>::from_floats([0.485, 0.456, 0.406], &device)
        .reshape([1, 3, 1, 1]);

let std = Tensor::<B, 1>::from_data([0.229, 0.224, 0.225], &device)
    .reshape([1, 3, 1, 1]);
```

and you can simply compute `(x - mean) / std`. Since this is part of the metric config, we could potentially have optional normalizer field instead of a bool, e.g.

```rust
pub struct Dists<B: Backend> {
    // ...
    normalize: Option<Normalizer<B>>,
}

pub struct Normalizer<B: Backend> {
    pub mean: Tensor<B, 4>,
    pub std: Tensor<B, 4>,
}
```

so the mean/std tensors are initialized with the metric.

Comment (koreaygj) [reply crates/burn-train/src/metric/vision/dists/metric.rs]: Addressed your feedback 👍🏻
[8683bf5](https://github.com/tracel-ai/burn/pull/4574/commits/8683bf568e3bf29fb559619b2d0fcaa144bc2564)
